### PR TITLE
dragonball: Introduce vhost-net device

### PR DESCRIPTION
**Pull request by justxuewei**
PLEASE NOTE THAT this pull request just implements vhost-net support for
Dragonball, and adaptation for the Runtime-rs. And this pull request
DOESN'T provide an item to config which backend to use. To sum up,
virtio-net as a default backend is only choice for the user so far.

This pull request introduces vhost-net device for the Dragonball. In
addition, this pull request includes changes of Runtime-rs to improve
network configuration abilities.

The Dragonball part implements a vhost-net device and a vhost-net device
manager, named `VhostNetDeviceMgr`, to manage vhost-net device.
`NetworkInterfaceConfig` is introduced as a high-level abstract for network
config. Then, the Dragonball is able to distinguish network backends, e.g.
virtio-net, vhost-net, vhost-user-net(WIP), etc.

The Runtime-rs part adds support of multiple network backends as well.
`NetworkConfig` has a couple of new fields, like `backend`,
`use_shared_irq`, etc. And Dragonball's network config structs are
implmented `From` trait which allow to be converted from the Runtime-rs's
network config conveniently.

Fixes: #7674

The functionality was tested by the iperf3.

test1: Vm as server, host as client

```
 -----------------------------------------------------------
 Server listening on 7575 (test #1)
 -----------------------------------------------------------
 Accepted connection from 192.168.155.1, port 45408
 [   45.877436] random: crng init done
 [  5] local 192.168.155.4 port 7575 connected to 192.168.155.1 port 45420
 [ ID] Interval           Transfer     Bitrate
 [  5]   0.00-1.00   sec  1.18 GBytes  10.1 Gbits/sec
 [  5]   1.00-2.00   sec  1.24 GBytes  10.7 Gbits/sec
 [  5]   2.00-3.00   sec  1.23 GBytes  10.5 Gbits/sec
 [  5]   3.00-4.00   sec  1.22 GBytes  10.5 Gbits/sec
 [  5]   4.00-5.00   sec  1.21 GBytes  10.4 Gbits/sec
 [  5]   5.00-6.00   sec  1.22 GBytes  10.5 Gbits/sec
 [  5]   6.00-7.00   sec  1.21 GBytes  10.4 Gbits/sec
 [  5]   7.00-8.00   sec  1.24 GBytes  10.6 Gbits/sec
 [  5]   8.00-9.00   sec  1.21 GBytes  10.4 Gbits/sec
 [  5]   9.00-10.00  sec  1.23 GBytes  10.6 Gbits/sec
 - - - - - - - - - - - - - - - - - - - - - - - - -
 [ ID] Interval           Transfer     Bitrate
 [  5]   0.00-10.00  sec  12.2 GBytes  10.5 Gbits/sec                  receiver
```

test2: Vm as client, host as server
```
[   74.592272] random: crng init done
Connecting to host 192.168.155.1, port 7575
[  5] local 192.168.155.4 port 34624 connected to 192.168.155.1 port 7575
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  1.13 GBytes  9.70 Gbits/sec    0    707 KBytes
[  5]   1.00-2.00   sec  1.23 GBytes  10.5 Gbits/sec    0    806 KBytes
[  5]   2.00-3.00   sec  1.19 GBytes  10.2 Gbits/sec    0    495 KBytes
[  5]   3.00-4.00   sec  1.22 GBytes  10.5 Gbits/sec    0    959 KBytes
[  5]   4.00-5.00   sec  1.22 GBytes  10.4 Gbits/sec    0    563 KBytes
[  5]   5.00-6.00   sec  1.20 GBytes  10.3 Gbits/sec    0    919 KBytes
[  5]   6.00-7.00   sec  1.21 GBytes  10.4 Gbits/sec    0    481 KBytes
[  5]   7.00-8.00   sec  1.21 GBytes  10.4 Gbits/sec    0    868 KBytes
[  5]   8.00-9.00   sec  1.20 GBytes  10.3 Gbits/sec    0    738 KBytes
[  5]   9.00-10.00  sec  1.20 GBytes  10.3 Gbits/sec    0    800 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  12.0 GBytes  10.3 Gbits/sec    0             sender
[  5]   0.00-10.04  sec  12.0 GBytes  10.3 Gbits/sec            receiver
```

### CODE REVIEW

File: src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs [hunk 1]
```diff
@@ -28,25 +33,90 @@ impl fmt::Debug for Address {
...
 
-#[derive(Clone, Debug, Default)]
-pub struct NetworkConfig {
-    /// for detach, now it's default value 0.
-    pub index: u64,
+#[derive(Clone, Debug)]
+pub enum NetworkBackend {
+    Virtio(VirtioConfig),
+    Vhost(VirtioConfig),
+}
```
Comment (Apokleos) [inline src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs:40]: What's the difference between the two backends? After read the related code, I find dup codes.

**Review by Apokleos — COMMENTED** [review]
Nice work!
but with some comments .

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 2]
```diff
@@ -293,9 +302,20 @@ impl VmmService {
...
+            VmmAction::InsertNetworkDevice(config) => match config.backend {
+                Backend::Virtio(_) => {
+                    #[cfg(not(feature = "virtio-net"))]
+                    panic!("virtio-net feature is not enabled");
+                    #[cfg(feature = "virtio-net")]
+                    self.add_virtio_net_device(vmm, event_mgr, config.into())
+                }
+                Backend::Vhost(_) => {
+                    #[cfg(not(feature = "vhost-net"))]
+                    panic!("vhost-net feature is not enabled");
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/api/v1/vmm_action.rs:314]: The reason is that there are two return types if no `panic!` here. If vhost-net feature is enabled, a `VmmRequestResult` is returned. If not, return nothing. As a result, Rust compiler feels confused.

File: src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs (hunk 1 above)
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs:40]: The configs, used by virtio-net and vhost-net, are same. However, their backends are different, see `src/dragonball/src/api/v1/vmm_action.rs` line 305-318.

File: src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs [hunk 3]
```diff
@@ -8,6 +8,11 @@ use std::fmt;
 
 use anyhow::{Context, Result};
 use async_trait::async_trait;
+use dbs_utils::net::MacAddr as DragonballMacAddr;
+use dragonball::api::v1::{
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs:12]: Yeah, I agree with you. I duplicates a set of types, like Backend, NetworkInterfaceConfig, etc, in Runtime-rs. Then, there are some functions to help convert Runtime-rs's type to Dragonball's type, named `DragonballXxx`. Therefore, the Dragonball-related types should be imported here.

File: src/runtime-rs/crates/hypervisor/src/dragonball/inner_device.rs [hunk 4]
```diff
@@ -207,25 +203,8 @@ impl DragonballInner {
...
-            num_queues: config.queue_num,
-            queue_size: config.queue_size as u16,
-            ..Default::default()
-        };
-
-        info!(
-            sl!(),
-            "add {} endpoint to {}", iface_cfg.host_dev_name, iface_cfg.iface_id
-        );
-
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/dragonball/inner_device.rs:226]: Fixed, see c48b238d0f37c7feb8efddba0e23fc90d245cc90.

File: src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs]: Fixed, see https://github.com/kata-containers/kata-containers/commit/c48b238d0f37c7feb8efddba0e23fc90d245cc90.

File: src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs (hunk 1 above)
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/device/driver/virtio_net.rs]: Fixed, see 301551eee3719f43d6fccb6126575f43e6aef57e.

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 5]
```diff
@@ -624,6 +644,30 @@ impl VmmService {
...
+        let ctx = vm
+            .create_device_op_context(Some(event_mgr.epoll_manager()))
+            .map_err(|err| {
+                if let StartMicroVmError::MicroVMAlreadyRunning = err {
+                    VmmActionError::VhostNet(VhostNetDeviceError::UpdateNotAllowedPostBoot)
+                } else if let StartMicroVmError::UpcallServerNotReady = err {
+                    VmmActionError::UpcallServerNotReady
+                } else {
+                    VmmActionError::StartMicroVm(err)
+                }
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/api/v1/vmm_action.rs]: Fixed, see 3baa3c434a42b534f0e41d73b057283b0a788902.

**Comment by justxuewei** [comment]
> Can commit message be summarized briefly? It seems that test information does not need to be placed in it.

I've push forcely a new commit which is removed those test information, and added more details.

@Apokleos Thanks for your code review. I left replies following your comments. PTAL, thanks!

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 6]
```diff
@@ -33,6 +33,10 @@ pub use crate::device_manager::fs_dev_mgr::{
 };
 #[cfg(feature = "virtio-mem")]
 pub use crate::device_manager::mem_dev_mgr::{MemDeviceConfigInfo, MemDeviceError};
+#[cfg(feature = "vhost-net")]
+use crate::device_manager::vhost_net_dev_mgr::VhostNetDeviceError;
+#[cfg(feature = "vhost-net")]
+pub use crate::device_manager::vhost_net_dev_mgr::{VhostNetDeviceConfigInfo, VhostNetDeviceMgr};
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/api/v1/vmm_action.rs]: Fixed, see https://github.com/kata-containers/kata-containers/commit/3baa3c434a42b534f0e41d73b057283b0a788902.

**Review by Apokleos — APPROVED** [review]
well done @justxuewei 
LGTM !

File: src/dragonball/Cargo.toml [hunk 7]
```diff
@@ -61,3 +61,4 @@ virtio-net = ["dbs-virtio-devices/virtio-net", "virtio-queue"]
 virtio-fs = ["dbs-virtio-devices/virtio-fs", "virtio-queue", "atomic-guest-memory"]
 virtio-mem = ["dbs-virtio-devices/virtio-mem", "virtio-queue", "atomic-guest-memory"]
 virtio-balloon = ["dbs-virtio-devices/virtio-balloon", "virtio-queue"]
+vhost-net = ["virtio-net", "dbs-virtio-devices/vhost-net"]
```
Comment (ZizhengBian) [inline src/dragonball/Cargo.toml:64]: Why vhost-net depends on virtio-net

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 8]
```diff
@@ -201,15 +204,20 @@ pub enum Error {
...
     VirtioFs(fs::Error),
 
+    #[cfg(feature = "vhost")]
+    #[error("vhost error: {0}")]
+    /// Error from vhost-net.
+    VhostNet(#[from] vhost_rs::Error),
+
+    #[cfg(feature = "virtio-net")]
+    #[error("tap device operation error: {0}")]
+    TapDeviceError(#[source] TapError),
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:219]: Why remove VirtioNetError? I think we should distinguish errors from VirtioNet and VhostNet

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+            ))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Device manager to manage all vhost net devices.
+pub struct VhostNetDeviceMgr {
+    pub(crate) info_list: DeviceConfigInfos<VhostNetDeviceConfigInfo>,
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:141]: Should these fields pub(crate)

File: src/runtime-rs/crates/hypervisor/src/dragonball/mod.rs [hunk 10]
```diff
@@ -187,3 +192,40 @@ impl Persist for Dragonball {
...
+        r.into()
+    }
+}
+
+impl From<&NetworkConfig> for DragonballNetworkConfig {
+    fn from(value: &NetworkConfig) -> Self {
+        let virtio_config = DragonballVirtioConfig {
+            iface_id: value.virt_iface_name.clone(),
+            host_dev_name: value.host_dev_name.clone(),
+            // TODO(justxuewei): rx_rate_limiter is not supported.
```
Comment (ZizhengBian) [inline src/runtime-rs/crates/hypervisor/src/dragonball/mod.rs:208]: Open an issue, and link with todo

File: src/dragonball/Cargo.toml (hunk 7 above)
Comment (justxuewei) [reply src/dragonball/Cargo.toml:64]: Fixed. Dragonball's vhost-net feature not depends on virtio-net feature now.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 8 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:219]: Fixed: Remove `TapError` from dbs-virtio-devices's Error, and add `VirtioNet` and `VhostNet` two fields.

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs (hunk 9 above)
Comment (justxuewei) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:141]: Fixed.

File: src/runtime-rs/crates/hypervisor/src/dragonball/mod.rs (hunk 10 above)
Comment (justxuewei) [reply src/runtime-rs/crates/hypervisor/src/dragonball/mod.rs]: Fixed.

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+
+use crate::{
+    address_space_manager::{GuestAddressSpaceImpl, GuestRegionImpl},
+    config_manager::{ConfigItem, DeviceConfigInfos},
+};
+
+use super::{DeviceManager, DeviceMgrError, DeviceOpContext};
+
+/// Default number of virtio queues, one rx/tx pair.
+pub const NUM_QUEUES: usize = 2;
```
Comment (studychao) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:20]: maybe `DEFAULT_NUM_QUEUES` makes more sense?

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 12]
```diff
@@ -0,0 +1,646 @@
...
+/// Default size of virtio queues.
+pub const QUEUE_SIZE: u16 = 256;
```
Comment (studychao) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:22]: maybe DEFAULT_QUEUE_SIZE makes more sense?

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 13]
```diff
@@ -292,10 +301,21 @@ impl VmmService {
...
             }
-            #[cfg(feature = "virtio-net")]
-            VmmAction::InsertNetworkDevice(virtio_net_cfg) => {
-                self.add_virtio_net_device(vmm, event_mgr, virtio_net_cfg)
-            }
+            #[cfg(any(feature = "virtio-net", feature = "vhost-net"))]
+            VmmAction::InsertNetworkDevice(config) => match config.backend {
+                Backend::Virtio(_) => {
+                    #[cfg(not(feature = "virtio-net"))]
+                    panic!("virtio-net feature is not enabled");
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/api/v1/vmm_action.rs]: Fixed, I've moved annotation of conditional compilation outside of body. Those panic statements are able to removed.

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs (hunk 11 above)
Comment (justxuewei) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs]: Fixed.

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs (hunk 12 above)
Comment (justxuewei) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs]: Fixed.

**Review by lifupan — APPROVED** [review]

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 14]
```diff
@@ -201,15 +205,20 @@ pub enum Error {
...
-    VirtioNetError(#[from] crate::net::NetError),
-
     #[cfg(feature = "virtio-fs")]
     /// Error from Virtio fs.
     #[error("virtio-fs error: {0}")]
     VirtioFs(fs::Error),
 
+    #[cfg(feature = "virtio-net")]
+    #[error("virtio-net error: {0}")]
+    VirtioNet(#[from] NetError),
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:215]: should we change transparency error transform to explicit error transform

Comment (ZizhengBian) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:215]: same below

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 15]
```diff
@@ -201,15 +205,20 @@ pub enum Error {
...
-    #[error("Virtio-net error: {0}")]
-    VirtioNetError(#[from] crate::net::NetError),
-
     #[cfg(feature = "virtio-fs")]
     /// Error from Virtio fs.
     #[error("virtio-fs error: {0}")]
     VirtioFs(fs::Error),
 
+    #[cfg(feature = "virtio-net")]
+    #[error("virtio-net error: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:214]: use {0:?} to print internal errors, same below

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 16]
```diff
@@ -219,6 +228,26 @@ pub enum Error {
...
 }
 
+// Error for tap devices
+#[cfg(feature = "virtio-net")]
+#[derive(Debug, thiserror::Error)]
+pub enum TapError {
+    #[error("missing {0} flags")]
+    MissingFlags(String),
+
+    #[error("failed to set offload: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:238]: use {0:?} to print internal errors, same below

File: src/dragonball/src/dbs_virtio_devices/src/net.rs [hunk 17]
```diff
@@ -60,15 +60,8 @@ pub const NET_EVENTS_COUNT: u32 = 6;
...
-    /// Open tap device failed.
-    #[error("open tap device failed: {0}")]
-    TapOpen(#[source] dbs_utils::net::TapError),
-    /// Setting tap interface offload flags failed.
-    #[error("set tap device vnet header size failed: {0}")]
-    TapSetOffload(#[source] dbs_utils::net::TapError),
-    /// Setting vnet header size failed.
-    #[error("set tap device vnet header size failed: {0}")]
-    TapSetVnetHdrSize(#[source] dbs_utils::net::TapError),
+    #[error("tap device operation error: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/net.rs:63]: {0:?}

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs [hunk 18]
```diff
@@ -0,0 +1,935 @@
...
+const NET_DRIVER_NAME: &str = "vhost-net";
+// Epoll token for control queue
+const CTRL_SLOT: u32 = 0;
+// Control queue size
+const CTRL_QUEUE_SIZE: u16 = 64;
+
+/// Error for vhost-net devices to handle requests from guests.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("tap device operation error: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs:53]: {0:?}

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs [hunk 19]
```diff
@@ -0,0 +1,935 @@
...
+                    .into_iter()
+                    .map(|flag| *flag)
+                    .collect::<Vec<&str>>()
+                    .join(", "),
+            ),
+        )));
+    }
+
+    // Set offload flags to match the virtio features below.
+    tap.set_offload(
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs:121]: set offload by acked features

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs [hunk 20]
```diff
@@ -0,0 +1,935 @@
...
+
+            let userspace_addr = region
+                .get_host_address(MemoryRegionAddress(0))
+                .map_err(|_| VirtioError::InvalidGuestAddress(guest_phys_addr))?;
+
+            regions.push(VhostUserMemoryRegionInfo {
+                guest_phys_addr: guest_phys_addr.raw_value(),
+                memory_size: region.len(),
+                userspace_addr: userspace_addr as *const u8 as u64,
+                mmap_offset: 0u64,
```
Comment (ZizhengBian) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs:357]: can we use type inference here？

File: src/dragonball/src/device_manager/mod.rs [hunk 21]
```diff
@@ -537,6 +543,9 @@ pub struct DeviceManager {
 
     #[cfg(feature = "virtio-balloon")]
     pub(crate) balloon_manager: BalloonDeviceMgr,
+
+    #[cfg(feature = "vhost-net")]
+    pub(crate) vhost_net_manager: VhostNetDeviceMgr,
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/mod.rs:548]: should we make these manager pub(crate)?

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 22]
```diff
@@ -0,0 +1,646 @@
+// Copyright (C) 2019-2023 Alibaba Cloud. All rights reserved.
+// Copyright (C) 2019-2023 Ant Group. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use dbs_utils::net::MacAddr;
+use dbs_virtio_devices::{vhost::vhost_kern::net::Net, Error as VirtioError};
+use serde::{Deserialize, Serialize};
+use std::{result::Result, sync::Arc};
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:9]: move std use statement to the top of all use statements

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 23]
```diff
@@ -0,0 +1,646 @@
...
+    #[error("the host device name {0} is already in use")]
+    HostDeviceNameInUse(String),
+    /// The update isn't allowed after booting the mircovm.
+    #[error("update operation is not allowed after booting")]
+    UpdateNotAllowedPostBoot,
+    /// Invalid queue number for vhost-net device.
+    #[error("invalid queue number {0} for vhost-net device")]
+    InvalidQueueNum(usize),
+    /// Failure from device manager.
+    #[error("failure in device manager operations: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:47]: {0:?}

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 24]
```diff
@@ -0,0 +1,646 @@
...
+    DeviceManager(#[source] DeviceMgrError),
+    /// Failure from virtio subsystem.
+    #[error(transparent)]
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:50]: It's better to leave a intact error stack

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 25]
```diff
@@ -0,0 +1,646 @@
...
+        ctx: &mut DeviceOpContext,
+    ) -> Result<Box<Net<GuestAddressSpaceImpl, QueueSync, GuestRegionImpl>>, VirtioError> {
+        slog::info!(
+            ctx.logger(),
+            "create a vhost-net device";
+            "subsystem" => "vhost_net_dev_mgr",
+            "id" => &cfg.iface_id,
+            "host_dev_name" => &cfg.host_dev_name,
+        );
+        let epoll_mgr = ctx.epoll_mgr.clone().ok_or(VirtioError::InvalidInput)?;
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:158]: why we use invalidinput error here?

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs [hunk 26]
```diff
@@ -0,0 +1,646 @@
...
+                "host_dev_name" => &info.config.host_dev_name,
+            );
+
+            let device = Self::create_device(&info.config, ctx)
+                .map_err(VhostNetDeviceError::CreateNetDevice)?;
+            let mmio_dev = DeviceManager::create_mmio_virtio_device(
+                device,
+                ctx,
+                info.config.use_shared_irq.unwrap_or(self.use_shared_irq),
+                info.config.use_generic_irq.unwrap_or(USE_GENERIC_IRQ),
```
Comment (ZizhengBian) [inline src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:244]: we don't have generic irq now, should we add this?

File: src/dragonball/src/error.rs [hunk 27]
```diff
@@ -193,6 +193,11 @@ pub enum StartMicroVmError {
     /// Virtio-balloon errors.
     #[error("virtio-balloon errors: {0}")]
     BalloonDeviceError(#[source] device_manager::balloon_dev_mgr::BalloonDeviceError),
+
+    /// Vhost-net device errors.
+    #[cfg(feature = "vhost-net")]
+    #[error("vhost-net errors: {0}")]
```
Comment (ZizhengBian) [inline src/dragonball/src/error.rs:199]: {0:?}

File: src/dragonball/src/api/v1/virtio_net.rs [hunk 28]
```diff
@@ -0,0 +1,217 @@
...
+    /// Allow duplicate mac
+    pub allow_duplicate_mac: bool,
+}
+
+/// This struct represents the strongly typed equivalent of the json body from
+/// net iface related requests.
+/// This struct works with virtio-net devices and vhost-net devices.
+#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize, Default)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkInterfaceConfig {
```
Comment (ZizhengBian) [inline src/dragonball/src/api/v1/virtio_net.rs:57]: Do we really need the convert from NetworkInterfaceConfig to NetworkInterfaceConfigInfo?

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs (hunk 26 above)
Comment (studychao) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:244]: I'll remove all generic irq from Dragonball in the later PR.

Comment (studychao) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:244]: see #8342

**Comment by studychao** [comment]
#8380 is added so that we don't forget to add that support in runtime-rs.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 14 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs (hunk 19 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_kern/net.rs]: Fixed

File: src/dragonball/src/device_manager/vhost_net_dev_mgr.rs (hunk 25 above)
Comment (ZizhengBian) [reply src/dragonball/src/device_manager/vhost_net_dev_mgr.rs:159]: Ok, we can deal with it later.

File: src/dragonball/src/api/v1/virtio_net.rs (hunk 28 above)
Comment (ZizhengBian) [reply src/dragonball/src/api/v1/virtio_net.rs:57]: Well, we can do this later.

**Review by ZizhengBian — APPROVED** [review]
LGTM

**Pull request merged**